E2E module upgrade test (#298)
kyma-project/btp-manager: docs/e2e_tests.md

## Overview

End-to-end (E2E) tests currently check if you can install, upgrade and uninstall BTP Manager using an OCI module image.
There are two separate tests:
- `e2e-tests` for checking installation and uninstallation of a given BTP Manager version
- `e2e-upgrade-tests` for checking BTP Manager upgradability from one version to another

> The flows of tests are mostly similar, upgrade tests contain extra steps for checking whether BTP Manager works as expected after upgrading to a new version. You can see the differences between the tests in the descriptions of the tests flows below.

Reviewer: ```suggestion
The flows of the tests are similar. The upgrade tests contain extra steps for checking whether BTP Manager works as expected after upgrading to a new version. You can see the differences between the tests in the descriptions of the tests' flows below.
```

10. Verify if BTP Operator has the `Ready` status equal to `True`.
11. Create a Service Instance and Service Binding with either real or dummy credentials. 
12. When real credentials are used, verify if the Service Instance and Service Binding have the `Ready` status `True`. If dummy credentials are used, verify if the `Ready` status for both of them is`NotProvisioned`.

> 13. Try to uninstall BTP Operator without force delete label.

Reviewer: Why are you saying `Try to uninstall...` and not `Uninstall..` 
In case of `Try to...`, I'd expect further instructions such as `If you succeed ...` and `If you don't succeed...` WDYT?

> 13. Try to uninstall BTP Operator without force delete label.

Reviewer: ```suggestion
13. Try to uninstall BTP Operator without the `force delete` label.
```

> 14. Verify if deprovisioning safety measures work.

Reviewer: ```suggestion
14. Verify if the deprovisioning safety measures work.
```

> 15. Add force delete label to BtpOperator CR.

Reviewer: ```suggestion
15. Add the `force delete` label to BtpOperator CR.
```

16. Verify if BTP Operator, ServiceInstance CRD and ServiceBinding CRD were deleted.
17. Uninstall BTP Manager. 

#### E2E tests for upgradability flow:
1. Create an OCI module image.
2. Push the image to the registry.
3. Create a Kubernetes cluster.
4. Wait for the new btp-operator OCI module image to be available in the registry.
5. Wait for the new btp-manager image to be available in the registry.
6. Download the latest btp-operator OCI module image.
7. Install the latest release of BTP Manager using `kubectl apply`.
8. Verify if deployment is in the `Available` state.
9. Install BTP Operator.
10. Verify if BTP Operator has the `Ready` status equal to `True`.
11. Create a Service Instance and Service Binding with real credentials.
12. Verify if the Service Instance and Service Binding have the `Ready` status `True`.
13. Download the new btp-operator OCI module image.
14. Upgrade BTP Manager to the new version using `kubectl apply`.
15. Verify if deployment is in the `Available` state.
16. Verify if the existing Service Instance and Service Binding have the `Ready` status `True`.
17. Create a new Service Binding with real credentials.

> 18. Verify if the new Service Binding have the `Ready` status `True`.

Reviewer: ```suggestion
18. Verify if the new Service Binding has the `Ready` status `True`.
```

> 19. Try to uninstall BTP Operator without force delete label.

Reviewer: ```suggestion
19. Try to uninstall BTP Operator without the `force delete` label.
```

> 20. Verify if deprovisioning safety measures work.

Reviewer: ```suggestion
20. Verify if the deprovisioning safety measures work.
```

> 21. Add force delete label to BtpOperator CR.

Reviewer: ```suggestion
21. Add the `force delete` label to BtpOperator CR.
```

22. Verify if BTP Operator, ServiceInstance CRD and ServiceBinding CRD were deleted.
23. Uninstall BTP Manager.

### CI pipelines
The Prow presubmit job, `pull-btp-manager-module-build`, creates the OCI module image. The `./scripts/create_module_image.sh` script does the actual execution.
This script sets appropriate environment variables and invokes `make module-build`. In effect, the module is built, and the OCI module image is pushed to the registry. 
The registry URL and component name are predefined. 

> **NOTE:**

> > For PR workflow runs the module image tag has the form `0.0.0-PR-<PR number>` due to component description requirements imposed by the tooling used.

Reviewer: ```suggestion
> For PR workflow runs, the module image tag has the form `0.0.0-PR-<PR number>` due to component description requirements imposed by the tooling used.
```

 
The GitHub Actions workflows execute the tests (`run-e2e-tests-reusable.yaml` and `run-e2e-upgrade-tests-reusable.yaml`). The Kubernetes cluster is created, and the sources are checked out.
The workflows wait till the OCI module image is available for fetching.
Scripts `./scripts/testing/run_e2e_module_tests.sh` and `./scripts/testing/run_e2e_module_upgrade_tests.sh` fetch the OCI module image from the registry. The scripts create the required prerequisites, 

> get the BTP Manager and BTP Operator installed or upgraded, validate expected statuses and the gets BTP Operator and BTP Manager uninstalled.

Reviewer: ```suggestion
get the BTP Manager and BTP Operator installed or upgraded, validate expected statuses, and get BTP Operator and BTP Manager uninstalled.
```

### Run E2E tests locally on k3d cluster
> **NOTE:**

> > Valid only for installing and uninstalling e2e tests.

Reviewer: ```suggestion
> Valid only for the installation and uninstallation e2e tests.
```
